appliance/postgresql2/simulator/simulator.go

```diff
@@ -0,0 +1,804 @@
+		Postgres: s.peer.Postgres.Info(),
+	})
+}
+
+func (s *Simulator) SetClusterState(args []string) {
+	if len(args) == 0 {
+		s.log.Error("missing state")
+		return
+	}
+	cs := s.discoverd.ClusterState()
+	if err := json.Unmarshal([]byte(args[0]), &cs.State); err != nil {
+		s.log.Error("error decoding state", "err", err)
+		return
+	}
+	s.discoverd.SetClusterState(cs)
+}
+
+func md5sum(data string) string {
+	digest := md5.Sum([]byte(data))
+	return hex.EncodeToString(digest[:])
+}
+
+type discoverdSimulator struct {
+	sync.Mutex
+	log     log15.Logger
+	state   *state.DiscoverdState
+	peers   []*discoverd.Instance
+	clients []*discoverdSimulatorClient
+}
+
+func newDiscoverdSimulator(log log15.Logger) *discoverdSimulator {
+	return &discoverdSimulator{
+		log:   log,
+		state: &state.DiscoverdState{},
+	}
+}
+
+func (d *discoverdSimulator) NewClient(inst *discoverd.Instance) *discoverdSimulatorClient {
+	c := &discoverdSimulatorClient{
+		d:      d,
+		inst:   inst,
+		events: make(chan *state.DiscoverdEvent),
+	}
+	d.clients = append(d.clients, c)
+	return c
+}
+
+func (d *discoverdSimulator) PeerJoined(inst *discoverd.Instance) {
+	d.Lock()
+	defer d.Unlock()
+
+	for _, p := range d.peers {
+		if p.ID == inst.ID {
+			d.log.Error("peer already exists", "addr", inst.Addr)
+			return
+		}
+	}
+
+	inst.Index = 1
+	if len(d.peers) > 0 {
+		inst.Index = d.peers[len(d.peers)-1].Index + 1
+	}
+	inst = inst.Clone()
+	d.peers = append(d.peers, inst)
+	for _, c := range d.clients {
+		go c.notifyPeersChanged()
+	}
+}
+
+func (d *discoverdSimulator) PeerRemoved(name string) {
+	d.Lock()
+	defer d.Unlock()
+
+	removed := false
+	for i, p := range d.peers {
+		if p.Meta["name"] == name {
+			d.peers = append(d.peers[:i], d.peers[i+1:]...)
+			removed = true
+			break
+		}
+	}
+	if !removed {
+		d.log.Error("peer not present", "name", name)
+		return
+	}
+	for _, c := range d.clients {
+		go c.notifyPeersChanged()
+	}
+}
+
+func (d *discoverdSimulator) SetClusterState(s *state.DiscoverdState) {
+	d.Lock()
+	defer d.Unlock()
+
+	if d.state.Index != s.Index {
+		panic(fmt.Sprintf("incorrect state index, have %d, want %d", s.Index, d.state.Index))
+	}
+	d.state.Index++
+	d.state.State = s.State.Clone()
+	s.Index = d.state.Index
+	for _, c := range d.clients {
+		go c.notifyStateChanged()
+	}
+}
+
+func (d *discoverdSimulator) ClusterState() *state.DiscoverdState {
+	d.Lock()
+	defer d.Unlock()
+
+	return &state.DiscoverdState{
+		Index: d.state.Index,
+		State: d.state.State.Clone(),
+	}
+}
+
+func (d *discoverdSimulator) Peers() []*discoverd.Instance {
+	d.Lock()
+	defer d.Unlock()
+
+	res := make([]*discoverd.Instance, len(d.peers))
+	copy(res, d.peers)
+	return res
+}
+
+type discoverdSimulatorClient struct {
+	sync.RWMutex
+	d      *discoverdSimulator
+	inst   *discoverd.Instance
+	events chan *state.DiscoverdEvent
+	init   bool
+}
+
+func (d *discoverdSimulatorClient) SetState(s *state.DiscoverdState) error {
+	time.Sleep(opLag)
```

Comment:
yeah, with the thought that it could help the race detector find interesting things.